### ISSUE TITLE
Composer 2 support

### ISSUE THREAD

**Issue by dan2k3k4** [amazeeio/lagoon#2257]
Currently [composer](http://getcomposer.org/) is added into the [`amazeeio/php` Dockerfile image](https://github.com/amazeeio/lagoon/blob/0e51f7b74d0b24473ca21d22fb7c0a1aa3029d8d/images/php/cli/Dockerfile#L10-L38) at version `1.10.9`.

We can update composer to the latest release by adding the following before we do the `RUN composer install` line in our own CLI Dockerfile for each project:

`RUN composer self-update --preview`

This will update to the latest preview release, which is currently `2.0.0-RC2`

There is no specific date yet for when Composer v2 will be out, however the [milestone is at 92%](https://github.com/composer/composer/milestone/23).

There is documentation on the upgrade step from [v1 to v2](https://getcomposer.org/upgrade/UPGRADE-2.0.md).

Eventually the Docker Hub image for `composer` will be updated to use v2 so our health checks will fail as we use [FROM composer:latest](https://github.com/amazeeio/lagoon/blob/0e51f7b74d0b24473ca21d22fb7c0a1aa3029d8d/images/php/fpm/Dockerfile#L7).

Do we need to implement a plan to support v2 across Lagoon and then allow people to do a `RUN composer self-update --rollback` to rollback to v1 for legacy?

Do we want to document how people can already upgrade to v2 for their own projects by adding `RUN composer self-update --preview`?

**Comment by dasrecht** [comment]
What we could do is co-install the composer versions and as soon as there's a sunset date for composer1 we'll basically switch them over after a grace period

**Comment by mxr576** [comment]
This is how we solved the "Composer 2 problem" in our dev stack.

https://github.com/Pronovix/docker-drupal-dev/commit/ba5d2cd4e6683010361ea39548484d749bec6ee1

This gives time to every developer to update those 2-5 usual Composer plugins on each project to a Composer 2 compatible version. Developers can easily switch between 1 and 2 via env vars.

**Comment by mxr576** [comment]
This probably makes supporting Composer 2 more urgent.

https://github.com/composer/composer/issues/9474

**Comment by Seldaek** [comment]
Please make at least composer2 available so users can easily opt-in. Composer 1 is kinda EOL as it is. We will do critical fixes if needed but that's about it. Thankfully it's pretty stable so should not need a ton of maintenance, but people should upgrade as v2 works so much better/faster.

I'd prefer a composer/composer1 approach where people have to opt-in to keep using v1 instead, at least for new projects.. But I'm not sure if that's feasible here.

**Comment by dasrecht** [comment]
Oh hai @Seldaek 👋 thanks for weighting in!

Our current goal is is to support Composer 1 as default (as we need to ensure backwards compatability for now) and add composer2 in one of the next releases.

You mention that Composer 1 is EOL - Do you have a date published where you discourage usage of Composer 1, similar to PHP's EOL - We start to adhere to those dates closer and closer to also help speeding up adoption.

**Comment by Seldaek** [comment]
We started warning on update a few weeks ago for versions older than `1.10.10` that they should upgrade to latest, and mentioning composer 2 is available.

We're ramping up usage (currently 22% of updates are done with v2) and quite aggressively fixing reported bugs, we haven't seen any major bugs really, mostly it's edge case stuff that wasn't caught in RCs. So I don't know when exactly we'll start warning *all* v1 users to update, but I'd guess by end of 2020 latest, because it's looking really stable. The only challenge remaining is the plugins, and that's not really something we can fix.

**Comment by dan2k3k4** [comment]
As Lagoon is hosting a majority of Drupal based projects, we might want to help push them to update their Composer plugins they use such as [cweagans/composer-patches](https://github.com/cweagans/composer-patches) which has Composer 2 support since [1.7.0](https://github.com/cweagans/composer-patches/releases/tag/1.7.0)

There's the "[Allow Drupal to work with Composer 2
](https://www.drupal.org/project/drupal/issues/3126566)" issue which is closed (fixed), and also a documentation page: "[Preparing your site for Composer 2](https://www.drupal.org/docs/develop/using-composer/preparing-your-site-for-composer-2)".

Maybe we could inform all clients about the change?

**Comment by mxr576** [comment]
> , but I'd guess by end of 2020 latest, 

> Maybe we could inform all clients about the change?

That is inevitable and I believe forced update and deprecation should not happen in such short time-frame. We are in the middle of number, lots of IT ppl take longer holidays in December, which means we can assume they have only 1 months to make those necessary changes on existing projects that are needed for Composer 2 compatibility. 

I expected the actual EOL of Composer 2 to the end 2021 Q1 the soonest which seemed a reasonable date for everyone to upgrade. (This is what I also wrote to our dev stack's [Dockerfile](https://github.com/Pronovix/docker-drupal-dev/compare/2.10.0...2.11.0-alpha5#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R12)).

**Comment by Seldaek** [comment]
Note that it's not because we start warning people on update that they will all migrate instantaneously. I'm sure we'll see people running v1 still in years. So yeah I'd rather start warning early, because people don't react quick, and I don't want to still be maintaining v1 in a few years. I think until end of next year for critical fixes is reasonable, but the measure of what is considered critical will probably get stricter as time goes :)

**Comment by tobybellwood** [comment]
Yup - this is something we're actively looking at.  One of my concerns is that Composer will drop support for all EOL PHP versions in a minor release, so we're just going through an information-gathering exercise at the moment.  We will be advising all customers to run the `composer self-update --2` in their own dockerfiles.  And yes, in our experience, people don't react quickly (cough, PHP5.6 users)

**Comment by Seldaek** [comment]
Regarding us dropping some php support, our plan is to keep supporting composer 2.1 with bugfixes for a couple years probably, so regardless which php version you're on you'll be in a better place if you upgrade to 2.x than if you stick to the already more or less dead 1.x.

**Comment by mxr576** [comment]
any update on this?

**Comment by dan2k3k4** [comment]
@mxr576 If you want Composer 2 in your images add: `RUN composer self-update --2` before your `RUN composer install` lines in your `Dockerfile`s that use composer.

I think the switch from 1 to 2 for the base/default images will be a little later on.
